src/seal.rs

```diff
@@ -0,0 +1,277 @@
+    };
+    Ok(ciphertext_vec)
+}
+
+#[inline]
+fn seal_finalize(ctx: &SealContext) -> Result<Vec<u8>>
+{
+    let output_len = ctx.get_output_length(0)?;
+    let ctx = ctx.handle;
+    let mut ciphertext_vec: Vec<u8> = Vec::with_capacity(output_len);
+    let mut ciphertext_len = 0;
+    let ciphertext = ciphertext_vec.as_mut_ptr() as *mut c_char;
+    let r = unsafe {
+        lib::yaca_seal_finalize(ctx, ciphertext, &mut ciphertext_len)
+    };
+    conv::res_c_to_rs(r)?;
+    assert!(ciphertext_len <= output_len);
+    unsafe {
+        ciphertext_vec.set_len(ciphertext_len);
+    };
+    Ok(ciphertext_vec)
+}
+
+/// Context for open operations
+pub struct OpenContext {
+    handle: *mut c_void,
+}
+
+impl Drop for OpenContext {
+    fn drop(&mut self)
+    {
+        unsafe {
+            lib::yaca_context_destroy(self.handle)
+        }
+    }
+}
+
+impl Context for OpenContext {
+    fn get_handle(&self) -> *mut c_void
+    {
+        self.handle
+    }
+}
+
+impl ContextWithPadding for OpenContext {}
+impl ContextWithRc2Supported for OpenContext {}
+impl ContextWithXcmDecryptProperties for OpenContext {
+    fn set_property_ccm_aad(&self, ccm_aad: &[u8], input_len: usize) -> Result<()>
+    {
+        open_set_input_length(self, input_len)?;
+        crypto::context_set_property_multiple(self, types::Property::CcmAad, ccm_aad)
+    }
+}
+
+impl OpenContext {
+    /// Initializes an decryption context.
+    pub fn initialize(prv_key: &Key, algo: &EncryptAlgorithm, bcm: &BlockCipherMode,
+                      sym_key_length: &KeyLength, sym_key: &Key,
+                      iv: Option<&Key>) -> Result<OpenContext>
+    {
+        open_initialize(prv_key, algo, bcm, sym_key_length, sym_key, iv)
+    }
+    /// Decrypts chunk of the data.
+    pub fn update(&self, plaintext: &[u8]) -> Result<Vec<u8>>
+    {
+        open_update(self, plaintext)
+    }
+    /// Decrypts the final chunk of the data.
+    pub fn finalize(&self) -> Result<Vec<u8>>
+    {
+        open_finalize(&self)
+    }
+}
+
+#[inline]
+fn open_initialize(prv_key: &Key, algo: &EncryptAlgorithm, bcm: &BlockCipherMode,
+                   sym_key_length: &KeyLength, sym_key: &Key,
+                   iv: Option<&Key>) -> Result<OpenContext>
+{
+    let prv_key = key::get_handle(&prv_key);
+    let algo = conv::encrypt_rs_to_c(algo);
+    let bcm = conv::bcm_rs_to_c(bcm);
+    let sym_key_bit_len = conv::key_length_rs_to_c(&sym_key_length);
+    let sym_key = key::get_handle(sym_key);
+    let iv = match iv {
+        Some(i) => key::get_handle(&i),
+        None => ptr::null(),
+    };
+    let mut handle = ptr::null_mut();
+    let r = unsafe {
+        lib::yaca_open_initialize(&mut handle, prv_key, algo, bcm,
+                                  sym_key_bit_len, sym_key, iv)
+    };
+    conv::res_c_to_rs(r)?;
+    assert!(!handle.is_null());
+    Ok(OpenContext{handle})
+}
+
+#[inline]
+fn open_set_input_length(ctx: &OpenContext, input_len: usize) -> Result<()>
+{
```

Comment:
TODO added.
I don't understand some of your thoughts here in regard to the issue though. Is there anything important you feel like I need to know, or you just want me to work on dedupping the code like in other comments?

Reply:
Just shared a braindump in hope of being helpful. Sorry for not being suitably transparent, we're wired a bit differently so that's gonna happen again.

Honestly, I'd be content with leaving it as it is. A TODO's fine as well of course. Just listed macros/traits/closures as three possible approaches to code dedup, should you choose to go that way.